apps/src/lib/node/ledger/shell/process_proposal.rs

```diff
@@ -67,116 +87,183 @@ where
+                    .into(),
             },
-            Ok(result) => match result {
-                // If it is a raw transaction, we do no further validation
-                TxType::Raw(_) => TxResult {
-                    code: ErrorCodes::InvalidTx.into(),
-                    info: "Transaction rejected: Non-encrypted transactions \
-                           are not supported"
-                        .into(),
-                },
-                TxType::Protocol(protocol_tx) => match protocol_tx.tx {
-                    ProtocolTxType::EthereumEvents(_) => TxResult {
-                        code: ErrorCodes::Ok.into(),
-                        info: "Process Proposal accepted this transaction"
-                            .into(),
-                    },
-                    _ => TxResult {
-                        code: ErrorCodes::InvalidTx.into(),
-                        info: "Unsupported protocol transaction type".into(),
-                    },
-                },
-                TxType::Decrypted(tx) => match self.next_wrapper() {
-                    Some(wrapper) => {
-                        if wrapper.tx_hash != tx.hash_commitment() {
-                            TxResult {
-                                code: ErrorCodes::InvalidOrder.into(),
-                                info: "Process proposal rejected a decrypted \
-                                       transaction that violated the tx order \
-                                       determined in the previous block"
-                                    .into(),
-                            }
-                        } else if verify_decrypted_correctly(&tx, privkey) {
+            TxType::Protocol(protocol_tx) => match protocol_tx.tx {
+                ProtocolTxType::EthereumEvents(digest) => {
+                    *vote_ext_digest_num += 1;
+
+                    let extensions =
+                        digest.decompress(self.storage.last_height);
+                    let filtered_extensions = self
+                        .filter_invalid_vote_extensions_residuals(extensions);
+
+                    let mut voting_power = FractionalVotingPower::default();
+                    let total_power = {
+                        let epoch =
+                            self.storage.block.pred_epochs.get_epoch(
+                                BlockHeight(self.storage.last_height.0),
+                            );
+                        u64::from(self.get_total_voting_power(epoch))
+                    };
+
+                    if filtered_extensions.into_iter().all(|maybe_ext| {
+                        maybe_ext
+                            .map(|(power, _)| {
+                                voting_power += FractionalVotingPower::new(
+                                    u64::from(power),
+                                    total_power,
+                                )
+                                .unwrap_or_default();
```

Comment:
we should hard `.unwrap()`/`.expect()` here or otherwise match on the result of constructing the `FractionalVotingPower`. Although I'm not sure how we would handle the error (or if we should ever encounter it - the voting power values are coming from our storage so we should always be able to construct a valid `FractionalVotingPower`?)